browser/ntp_background_images/ntp_p3a_helper_impl.cc

```diff
@@ -0,0 +1,162 @@
+constexpr base::TimeDelta kLandingTime = base::Seconds(10);
+
+}  // namespace
+
+NTPP3AHelperImpl::NTPP3AHelperImpl(PrefService* local_state,
+                                   brave::BraveP3AService* p3a_service)
+    : local_state_(local_state), p3a_service_(p3a_service) {
+  DCHECK(local_state);
+  DCHECK(p3a_service);
+}
+
+void NTPP3AHelperImpl::RegisterLocalStatePrefs(PrefRegistrySimple* registry) {
+  registry->RegisterDictionaryPref(kNewTabPageEventCountDictPref);
+}
+
+void NTPP3AHelperImpl::RecordView(const std::string& creative_instance_id) {
+  UpdateMetricCount(creative_instance_id, kViewEventKey);
+}
+
+void NTPP3AHelperImpl::RecordClickAndMaybeLand(
+    const std::string& creative_instance_id) {
+  UpdateMetricCount(creative_instance_id, kClickEventKey);
+  landing_check_timer_.Start(
+      FROM_HERE, kStartLandingCheckTime,
+      base::BindOnce(&NTPP3AHelperImpl::OnLandingStartCheck,
+                     base::Unretained(this), creative_instance_id));
+}
+
+void NTPP3AHelperImpl::SetLastTabURL(const GURL& url) {
+  last_tab_hostname_ = url.host();
+}
+
+std::string NTPP3AHelperImpl::GetHistogramName(
+    const std::string& creative_instance_id,
+    const std::string& event_type) {
+  std::string result = "creativeInstanceId." + creative_instance_id;
+  return base::StrCat(
+      {brave::kCreativeMetricPrefix, creative_instance_id, ".", event_type});
+}
+
+void NTPP3AHelperImpl::UpdateMetricCount(
+    const std::string& creative_instance_id,
+    const std::string& event_type) {
+  const std::string histogram_name =
+      GetHistogramName(creative_instance_id, event_type);
+  const base::Time histogram_expiry_time =
+      base::Time::Now() + kHistogramExpiryTime;
+
+  p3a_service_->RegisterMetric(histogram_name, brave::MetricLogType::kExpress,
+                               histogram_expiry_time);
+
+  DictionaryPrefUpdate update(local_state_, kNewTabPageEventCountDictPref);
+  int count = 0;
+
+  base::Value::Dict& update_dict = update->GetDict();
+
+  const std::string full_pref_key =
+      base::StrCat({creative_instance_id, ".", event_type});
+  const std::string full_expiry_key =
+      base::StrCat({creative_instance_id, ".", kExpireTimeKey});
+
+  const base::Value* current_value =
+      update_dict.FindByDottedPath(full_pref_key);
+  if (current_value && current_value->is_int()) {
+    count = current_value->GetInt();
+  }
+
+  count++;
+  update_dict.SetByDottedPath(full_pref_key, base::Value(count));
+  const base::Time new_expiry_time = base::Time::Now() + kCountExpiryTime;
+  update_dict.SetByDottedPath(full_expiry_key,
+                              base::TimeToValue(new_expiry_time));
+  p3a_utils::RecordToHistogramBucket(histogram_name.c_str(), kCountBuckets,
+                                     count);
+}
+
+void NTPP3AHelperImpl::DeleteOldCounts() {
+  const base::Value::Dict& dict =
+      local_state_->GetDictionary(kNewTabPageEventCountDictPref)->GetDict();
```

Comment:
GetDictionary is deprecated, please see https://source.chromium.org/chromium/chromium/src/+/main:base/values.h